src/librustc_const_eval/eval.rs

```diff
@@ -384,84 +395,83 @@ fn eval_const_expr_partial<'a, 'tcx>(cx: &ConstContext<'a, 'tcx>,
           };
           callee_cx.eval(&body.value)?
       },
-      hir::ExprLit(ref lit) => match lit_to_const(&lit.node, tcx, ety) {
-          Ok(val) => val,
+      hir::ExprLit(ref lit) => match lit_to_const(&lit.node, tcx, ty) {
+          Ok(val) => mk_const(val),
           Err(err) => signal!(e, err),
       },
       hir::ExprBlock(ref block) => {
         match block.expr {
             Some(ref expr) => cx.eval(expr)?,
-            None => Tuple(vec![]),
+            None => mk_const(Aggregate(Tuple(&[]))),
         }
       }
       hir::ExprType(ref e, _) => cx.eval(e)?,
       hir::ExprTup(ref fields) => {
-        Tuple(fields.iter().map(|e| cx.eval(e)).collect::<Result<_, _>>()?)
+        let values = fields.iter().map(|e| cx.eval(e)).collect::<Result<Vec<_>, _>>()?;
+        mk_const(Aggregate(Tuple(tcx.alloc_const_slice(&values))))
```

Comment:
Shouldn't we be using `mk_const_slice` here? That would avoid the `Vec` in the common(?) case of having less than eight fields.

There seem to be more cases of this in this file as well.